modules/doi/islandora_doi.install

```diff
@@ -0,0 +1,41 @@
+<?php
+
+/**
+ * @file
+ * Install and update hook implementations.
+ */
+
+/**
+ * Implements hook_uninstall().
+ */
+function islandora_doi_uninstall() {
+  $variables = array(
+    'islandora_doi_doiurl',
+    'islandora_doi_openurl',
+    'islandora_doi_openurl_pid',
+    'islandora_doi_use_legacy_mode',
+  );
+  array_walk($variables, 'variable_del');
+}
+
+/**
+ * Update variable defaults.
+ */
+function islandora_doi_update_7100() {
+  $variables = array(
+    'islandora_doi_openurl' => array(
+      'old_default' => 'http://www.crossref.org/openurl',
+      'new_default' => ISLANDORA_DOI_OPENURL_DEFAULT,
+    ),
+    'islandora_doi_openurl_pid' => array(
+      'old_default' => 'user@example.org',
+      'new_default' => ISLANDORA_DOI_OPENURL_PID_DEFAULT,
```

Comment:
Very minor, and can probably be ignored, but: Update hooks are a special thing... normally, it best to avoid using references to data outside of them, and have all values explicitly contained in the update function; for example, if we want to change the defaults again in the future, we would presumably roll another `hook_update_N()` to similarly update like we do here; however if an installation was to update across both hooks, the state when entering the second update hook may be unexpected... it would already contain the new values. This is much [more of an issue with database/schema management things](https://www.drupal.org/node/150220), but seems similar enough that I thought I should at least mention it.

Reply:
@adam-vessey so you mean, don't use the constant here, just put the new default manually, right?

Reply:
Yeah, both here and [above](https://github.com/Islandora/islandora_scholar/pull/285/files#diff-918824dec4ab3033cac4e6b3ac88b1daR28), to have them explicitly as strings instead of using the constants.

Reply:
Good point @adam-vessey! Thank you! @bryjbrown would you like to change this? The part with the default PID can then be taken out as it is superfluous anyway (I wrote it only for symmetry).

One question, though: if we were to put a second update function in here with the now current values as literals and the new values as constants (like in the first function), a user who already executed the first one would execute the second one just fine, right? But if they did not, the first execution would update the variables directly to the current constants (which is the concern here, correct?). But the second update hook would then not do anything, no? The retrieved value would be the `'new_default'`, and the `foreach` clause would not do any work. True, this is wasted computation time, but the result is the one expected. Or am I missing something here?

Reply:
No, you've got it @d-r-p... this is why I said it's very minor and can probably be ignored. The only way it would be an issue if some _other_ logic which depended on the existing value was being performed which is... entirely hypothetical. I can't really think of a case for it being necessary for this in particular, but this is not to say that none exists.

Reply:
@adam-vessey @bryjbrown @d-r-p i think i read somewhere (probably the hook_update_N page) that its good practice to remove old hook_update_N implementations from time to time (like keep the latest 2 old ones), which means, i guess, that we can not depend too much on them being an incremental fix, but be as standalone as possible. Right?

Reply:
@DiegoPino: Not... quite. From [`hook_update_N()`](https://api.drupal.org/api/drupal/modules!system!system.api.php/function/hook_update_N/7.x):

> A good rule of thumb is to remove updates older than two major releases of Drupal.

They're appear to be talking about major versions of _Drupal_, so if we had modules with update hooks from Drupal 4 era, we would/should have removed them for Drupal 7... or such is my understanding of this.

Reply:
hu! @adam-vessey++ true. Weird... can't think of anything 4 or 5 or 6 era that still works on 7! maybe they were talking about non-semantic versioning, where a major release could be any other subversion. (like 7.4 to 7.5?). Who knows..!

Reply:
Okay, so I was out yesterday and I'm just now getting to this and reading through. I'm not totally sure I understand whats going on here, if we change the "new_default" constant to its string literal value would it just be the exact same thing as "old_default"? For example, instead of 
```
  $variables = array(
    'islandora_doi_openurl' => array(
      'old_default' => 'http://www.crossref.org/openurl',
      'new_default' => ISLANDORA_DOI_OPENURL_DEFAULT,
    ),
    'islandora_doi_openurl_pid' => array(
      'old_default' => 'user@example.org',
      'new_default' => ISLANDORA_DOI_OPENURL_PID_DEFAULT,
    ),
  );
```
```
  $variables = array(
    'islandora_doi_openurl' => array(
      'old_default' => 'http://www.crossref.org/openurl',
      'new_default' => ISLANDORA_DOI_OPENURL_DEFAULT,
    ),
    'islandora_doi_openurl_pid' => array(
      'old_default' => 'user@example.org',
      'new_default' => ISLANDORA_DOI_OPENURL_PID_DEFAULT,
    ),
  );
```
we would instead have
```
  $variables = array(
    'islandora_doi_openurl' => array(
      'old_default' => 'http://www.crossref.org/openurl',
      'new_default' => ISLANDORA_DOI_OPENURL_DEFAULT,
    ),
    'islandora_doi_openurl_pid' => array(
      'old_default' => 'user@example.org',
      'new_default' => ISLANDORA_DOI_OPENURL_PID_DEFAULT,
    ),
  );
```
```
  $variables = array(
    'islandora_doi_openurl' => array(
      'old_default' => 'http://www.crossref.org/openurl',
      'new_default' => 'http://www.crossref.org/openurl',
    ),
    'islandora_doi_openurl_pid' => array(
      'old_default' => 'user@example.com',
      'new_default' => 'user@example.com',
    ),
  );
```
That seems like a pretty pointless update. If it sourced NEW values from constants it would make sense, but replacing them with string literals defeats that purpose.

Reply:
@bryjbrown so... old and new are the same? and why do we need the update hook then??? ❓ 😬

Reply:
@bryjbrown If the values are not changing, what's the point of this update hook at all, independent of whether the values are explicitly inline or referenced via a constant?... That said, it _does_ appear that the [`islandora_doi_openurl` value is being changed](https://github.com/Islandora/islandora_scholar/pull/285/files#diff-1c1e5251eca683f1ba22e7afb4f2a63cR8)... the TLD mixup which was fixed in  https://github.com/Islandora/islandora_scholar/pull/285/commits/fb4b7d6e7304cca1d0edf5be72d4e737085dcc35 was slightly confusing the issue...

Reply:
Thank you, @bryjbrown, for spotting the wrong tld!

Regarding your question: as far as I understood, the update hook implementation should read (although it could be simplified):
```php
/**
 * Update variable defaults.
 */
function islandora_doi_update_7100() {
  $variables = array(
    'islandora_doi_openurl' => array(
      'old_default' => 'http://www.crossref.org/openurl',
      'new_default' => 'https://www.crossref.org/openurl',
    ),
  );
  foreach ($variables as $variable => $info) {
    $value = variable_get($variable, $info['new_default']);
    if ($value == $info['old_default']) {
      variable_set($variable, $info['new_default']);
    }
  }
}
```
instead of:
```php
/**
 * Update variable defaults.
 */
function islandora_doi_update_7100() {
  $variables = array(
    'islandora_doi_openurl' => array(
      'old_default' => 'http://www.crossref.org/openurl',
      'new_default' => ISLANDORA_DOI_OPENURL_DEFAULT,
    ),
    'islandora_doi_openurl_pid' => array(
      'old_default' => 'user@example.com',
      'new_default' => ISLANDORA_DOI_OPENURL_PID_DEFAULT,
    ),
  );
  foreach ($variables as $variable => $info) {
    $value = variable_get($variable, $info['new_default']);
    if ($value == $info['old_default']) {
      variable_set($variable, $info['new_default']);
    }
  }
}
```
The purpose of this update is to use https instead of http when querying CrossRef. As I said earlier, I had included the OpenURL PID update only for symmetry.

Apologies for the tld mix-up from which the confusion seems to stem.

Reply:
@adam-vessey Yes, I think it was just a typo.

Reply:
@DiegoPino @adam-vessey So... :+1: to getting rid of this update hook?

@d-r-p Please chime in if we are misunderstanding the purpose of this update hook.

Reply:
@d-r-p Okay, didn't catch the s/http/https part on here, that is a valid update. I'll patch in your first example in an incoming commit. :+1:

Reply:
@bryjbrown no, we need the update hook, or users who set their OpenURL PID will continue to use http for querying CrossRef. I think what I wrote above should do it...

Reply:
@d-r-p Check https://github.com/Islandora/islandora_scholar/pull/285/commits/c9b85c976aa553de6078631f2e7456b5f8398776 to verify that this is what it should look like.

Reply:
Looks good to me! Thank you @bryjbrown!

@adam-vessey do you agree, as well?

Reply:
Yeah, looks good here.

Reply:
Awesome! I think we've finally driven a stake through this one's heart, @DiegoPino :fireworks: :champagne: